// @aws-crypto/kms-keyring-browser/build/main/src/kms_keyring_browser.js
"use strict";
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.KMS = exports.cacheClients = exports.excludeRegions = exports.limitRegions = exports.getKmsClient = exports.cacheKmsClients = exports.getClient = exports.KmsKeyringBrowser = void 0;
const kms_keyring_1 = require("@aws-crypto/kms-keyring");
Object.defineProperty(exports, "getClient", { enumerable: true, get: function () { return kms_keyring_1.getClient; } });
Object.defineProperty(exports, "limitRegions", { enumerable: true, get: function () { return kms_keyring_1.limitRegions; } });
Object.defineProperty(exports, "excludeRegions", { enumerable: true, get: function () { return kms_keyring_1.excludeRegions; } });
Object.defineProperty(exports, "cacheClients", { enumerable: true, get: function () { return kms_keyring_1.cacheClients; } });
const material_management_browser_1 = require("@aws-crypto/material-management-browser");
const aws_sdk_1 = require("aws-sdk");
Object.defineProperty(exports, "KMS", { enumerable: true, get: function () { return aws_sdk_1.KMS; } });
const version_1 = require("./version");
const getKmsClient = (0, kms_keyring_1.getClient)(aws_sdk_1.KMS, {
    customUserAgent: `AwsEncryptionSdkJavascriptBrowser/${version_1.version}`,
});
exports.getKmsClient = getKmsClient;
const cacheKmsClients = (0, kms_keyring_1.cacheClients)(getKmsClient);
exports.cacheKmsClients = cacheKmsClients;
class KmsKeyringBrowser extends (0, kms_keyring_1.KmsKeyringClass)(material_management_browser_1.KeyringWebCrypto) {
    constructor({ clientProvider = cacheKmsClients, keyIds, generatorKeyId, grantTokens, discovery, } = {}) {
        super({ clientProvider, keyIds, generatorKeyId, grantTokens, discovery });
    }
    async _onEncrypt(material) {
        const _material = await super._onEncrypt(material);
        return (0, material_management_browser_1.importForWebCryptoEncryptionMaterial)(_material);
    }
    async _onDecrypt(material, encryptedDataKeys) {
        const _material = await super._onDecrypt(material, encryptedDataKeys);
        return (0, material_management_browser_1.importForWebCryptoDecryptionMaterial)(_material);
    }
}
exports.KmsKeyringBrowser = KmsKeyringBrowser;
(0, material_management_browser_1.immutableClass)(KmsKeyringBrowser);
