// @aws-crypto/web-crypto-backend/build/main/src/index.js
"use strict";
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
Object.defineProperty(exports, "__esModule", { value: true });
exports.synchronousRandomValues = exports.isFullSupportWebCryptoBackend = exports.getZeroByteSubtle = exports.getNonZeroByteBackend = exports.configureFallback = exports.getWebCryptoBackend = void 0;
const util_locate_window_1 = require("@aws-sdk/util-locate-window");
const backend_factory_1 = require("./backend-factory");
const { getWebCryptoBackend, configureFallback } = (0, backend_factory_1.webCryptoBackendFactory)((0, util_locate_window_1.locateWindow)());
exports.getWebCryptoBackend = getWebCryptoBackend;
exports.configureFallback = configureFallback;
var backend_factory_2 = require("./backend-factory");
Object.defineProperty(exports, "getNonZeroByteBackend", { enumerable: true, get: function () { return backend_factory_2.getNonZeroByteBackend; } });
Object.defineProperty(exports, "getZeroByteSubtle", { enumerable: true, get: function () { return backend_factory_2.getZeroByteSubtle; } });
Object.defineProperty(exports, "isFullSupportWebCryptoBackend", { enumerable: true, get: function () { return backend_factory_2.isFullSupportWebCryptoBackend; } });
var synchronous_random_values_1 = require("./synchronous_random_values");
Object.defineProperty(exports, "synchronousRandomValues", { enumerable: true, get: function () { return synchronous_random_values_1.synchronousRandomValues; } });
