// @aws-crypto/raw-rsa-keyring-browser/build/module/src/types.js
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
export var RsaPadding;
(function (RsaPadding) {
    RsaPadding["OAEP_SHA1_MFG1"] = "OAEP_SHA1_MFG1";
    RsaPadding["OAEP_SHA256_MFG1"] = "OAEP_SHA256_MFG1";
    RsaPadding["OAEP_SHA384_MFG1"] = "OAEP_SHA384_MFG1";
    RsaPadding["OAEP_SHA512_MFG1"] = "OAEP_SHA512_MFG1";
})(RsaPadding || (RsaPadding = {}));
export var Format;
(function (Format) {
    Format["raw"] = "raw";
    Format["pkcs8"] = "pkcs8";
    Format["spki"] = "spki";
})(Format || (Format = {}));
// RSA_PKCS1, https://github.com/aws/aws-encryption-sdk-python/blob/master/src/aws_encryption_sdk/identifiers.py#L262
export var JsonWebKeyRsaAlg;
(function (JsonWebKeyRsaAlg) {
    JsonWebKeyRsaAlg["RSA-OAEP"] = "RSA-OAEP";
    JsonWebKeyRsaAlg["RSA-OAEP-256"] = "RSA-OAEP-256";
    JsonWebKeyRsaAlg["RSA-OAEP-384"] = "RSA-OAEP-384";
    JsonWebKeyRsaAlg["RSA-OAEP-512"] = "RSA-OAEP-512";
})(JsonWebKeyRsaAlg || (JsonWebKeyRsaAlg = {}));
export var RsaHash;
(function (RsaHash) {
    RsaHash["SHA-1"] = "SHA-1";
    RsaHash["SHA-256"] = "SHA-256";
    RsaHash["SHA-384"] = "SHA-384";
    RsaHash["SHA-512"] = "SHA-512";
})(RsaHash || (RsaHash = {}));
