// @aws-crypto/kms-keyring-browser/build/module/src/kms_keyring_browser.js
// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import { KmsKeyringClass, getClient, limitRegions, excludeRegions, cacheClients, } from '@aws-crypto/kms-keyring';
import { immutableClass, importForWebCryptoEncryptionMaterial, importForWebCryptoDecryptionMaterial, KeyringWebCrypto, } from '@aws-crypto/material-management-browser';
import { KMS } from 'aws-sdk';
import { version } from './version';
const getKmsClient = getClient(KMS, {
    customUserAgent: `AwsEncryptionSdkJavascriptBrowser/${version}`,
});
const cacheKmsClients = cacheClients(getKmsClient);
export class KmsKeyringBrowser extends KmsKeyringClass(KeyringWebCrypto) {
    constructor({ clientProvider = cacheKmsClients, keyIds, generatorKeyId, grantTokens, discovery, } = {}) {
        super({ clientProvider, keyIds, generatorKeyId, grantTokens, discovery });
    }
    async _onEncrypt(material) {
        const _material = await super._onEncrypt(material);
        return importForWebCryptoEncryptionMaterial(_material);
    }
    async _onDecrypt(material, encryptedDataKeys) {
        const _material = await super._onDecrypt(material, encryptedDataKeys);
        return importForWebCryptoDecryptionMaterial(_material);
    }
}
immutableClass(KmsKeyringBrowser);
export { getClient, cacheKmsClients, getKmsClient, limitRegions, excludeRegions, cacheClients, KMS, };
